README.md
# kings-ransom
kings ransom baseball model


logs/history.txt
Day 2
Marlins at Phillies: 15.2, 14.5: Marlins
Royals at Indians: 12, 13.8: Indians
Brewers at Cubs: 14, 13.2: Cubs
Orioles at Red Sox: 11.3, 14.5: Red Sox
Rockies at Rangers: 16, 12.3: Rangers
Twins at White Sox: 14.5, 14.2: Twins
Pirates at Cardinals: 14.2, 13.5: Cardinals
Mariners at Astros: 14.7, 15.3: Astros
D-backs at Padres: 12.6, 14.2: Padres
Angels at Athletics: 14.8, 15.1: Athletics
Giants at Dodgers: 15.2, 15.1: Dodgers

Day 3
Brewers at Cubs: 14.9, 14.1: Brewers
Orioles at Red Sox: 12.7, 14.9: Orioles
Twins at White Sox: 17.4, 14.2: White Sox
Pirates at Cardinals: 12.8, 14.7: Cardinals
Blue Jays at Rays: 14.9, 14.3: Rays
Marlins at Phillies: 13.3, 14.8: Phillies
Rockies at Rangers: 14.5, 12.9: Rockies
Braves at Mets: 14.6, 15.8: Braves
Angels at Athletics: 13.3, 13.2: Angels
Mariners at Astros: 20, 14.2: Astros
Giants at Dodgers: 16.3, 16.4: Giants
Royals at Indians: N/A: Royals
Tigers at Reds: 13.6, 16.5: Tigers
Yankees at Nationals: 16.3, 13.1: Nationals
D-backs at Padres: 14.6, 13.4: Padres

Day 4
Yankees at Nationals: 15.5, 13.8: Yankees
Marlins at Phillies: 12.9, 15.7: Marlins
Royals at Indians: 12.8, 15.5: Indians
Blue Jays at Rays: N/A: Rays
Tigers at Reds: 14, 14.6: Tigers
Orioles at Red Sox: 13.2, 13.4: Orioles
Twins at White Sox: 14.3, 14: Twins
Mariners at Astros: 13.5, 14.3: Mariners
Pirates at Cardinals: 13.8, 13.3: Pirates
Brewers at Cubs: 16.4, 13.5: Cubs
Rockies at Rangers: 14.4, 14.5: Rockies
Angels at Athletics: 14.5, 13.3: Athletics
D-backs at Padres: 13.9, 14: D-backs
Braves at Mets: 14.1, 16: Braves
Giants at Dodgers: 13.8, 15.5: Giants

Day 5
Angels at Athletics: 14.025, 15.82: Athletics
D-backs at Padres: 14.235, 14.453: Padres
Blue Jays at Nationals: 14.584, 13.855: Blue Jays
Cubs at Reds: 14.392, 15.283: Cubs
Braves at Rays: 15.183, 14.056: Rays
Royals at Tigers: 13.127, 13.644: Royals
Mariners at Astros: 14.353, 16.638: Astros
Mets at Red Sox: 16.078, 13.804: Mets
White Sox at Indians: 14.823, 15.592: PPD
Brewers at Pirates: 15.174, 14.331: Brewers

2020-07-28
White Sox at Indians: 14.909, 16.496: Indians
Blue Jays at Nationals: 14.468, 13.753: Blue Jays
Braves at Rays: 11.583, 17.292: Rays
Brewers at Pirates: 15.545, 15.732: Pirates
Royals at Tigers: N/A: Tigers
Mets at Red Sox: N/A: Mets
Cardinals at Twins: 13.293, 15.136: Twins
D-backs at Rangers: 15.329, 14.363: D-backs
Cubs at Reds: 12.856, 16.554: Cubs
Dodgers at Astros: 16.996, 13.175: Dodgers
Mariners at Angels: 14.453, 15.169: Angels
Rockies at Athletics: 14.119, 13.492: Rockies
Padres at Giants: 15.201, 12.953: Padres

2020-07-29
Rockies at Athletics: 13.373, 12.449: Rockies
D-backs at Rangers: 15.482, 12.571: Rangers
White Sox at Indians: 13.421, 17.113: White Sox
Royals at Tigers: 12.489, 12.456: Tigers
Cubs at Reds: 13.963, 14.513: Reds
Brewers at Pirates: 14.417, 15.202: Brewers
Dodgers at Astros: N/A: Dodgers
Cardinals at Twins: 12.285, 15.715: Twins
Red Sox at Mets: 14.512, 13.729: Red Sox
Nationals at Blue Jays: N/A: Nationals
Padres at Giants: 16.246, 12.927: Giants
Rays at Braves: 14.473, 14.26: Braves
Mariners at Angels: 11.197, 17.88: Mariners

2020-07-30
Indians at Twins: 15.428, 14.08: Indians
Red Sox at Mets: 13.94, 16.401: Red Sox
Rays at Braves: 14.953, 15.279: Braves
Royals at Tigers: N/A: Royals
Yankees at Orioles: N/A: Yankees
Mariners at Angels: 12.133, 15.088: Mariners
Dodgers at D-backs: 15.866, 14.414: Dodgers
Padres at Giants: 15.104, 13.143: Padres
Nationals at Blue Jays: N/A: Nationals

2020-07-31
Red Sox at Yankees: 14.407, 16.66: Yankees
Mets at Braves: 15.902, 13.643: Braves
Reds at Tigers: 14.988, 13.766: Reds
Rays at Orioles: 14.411, 12.369: Orioles
White Sox at Royals: N/A: White Sox
Indians at Twins: 14.327, 15.09: Twins
Padres at Rockies: 13.315, 14.343: Padres
Pirates at Cubs: 14.139, 14.459: Cubs
Astros at Angels: N/A: Astros
Rangers at Giants: 15.235, 16.763: Giants
Athletics at Mariners: 13.119, 13.265: Mariners
Dodgers at D-backs: 15.41, 15.861: D-backs

2020-08-01
Reds at Tigers: 15.499, 13.597: Reds
Astros at Angels: 17.296, 15.492: Angels
Indians at Twins: 14.632, 14.763: Twins
Red Sox at Yankees: 12.924, 18.172: Yankees
Pirates at Cubs: 14.566, 12.652: Cubs
White Sox at Royals: 14.995, 14.178: White Sox
Mets at Braves: 15.574, 13.995: Braves
Rays at Orioles: 14.548, 13.46: Orioles
Dodgers at D-backs: 14.334, 13.217: Dodgers
Padres at Rockies: 15.119, 16.426: Rockies
Rangers at Giants: 12.993, 12.951: Giants

2020-08-02
Reds at Tigers: N/A: Reds
Rays at Orioles: 14.917, 11.814: Orioles
Mets at Braves: N/A: Braves
White Sox at Royals: 12.723, 14.302: White Sox
Indians at Twins: 15.906, 17.203: Twins
Pirates at Cubs: 15.496, 13.28: Cubs
Padres at Rockies: 14.525, 16.229: Rockies
Rangers at Giants: 12.85, 13.961: Rangers
Astros at Angels: 14.025, 14.588: Astros
Dodgers at D-backs: 16.202, 12.168: Dodgers
Athletics at Mariners: 16.02, 10.152: Athletics
Red Sox at Yankees: 13.277, 15.205: Yankees

2020-08-03
Phillies at Yankees: 13.488, 14.958: Yankees
Mets at Braves: 15.466, 13.485: Mets
Indians at Reds: 15.013, 13.826: Reds
Pirates at Twins: 15.091, 17.694: Twins
White Sox at Brewers: 13.908, 14.458: White Sox
Royals at Cubs: 14.065, 12.891: Cubs
Giants at Rockies: 15.039, 15.243: Rockies
Athletics at Mariners: 13.158, 18.82: Athletics
Dodgers at Padres: 14.368, 13.51: Padres

2020-08-04
Pirates at Twins: 13.225, 13.887: Twins
Indians at Reds: 14.212, 14.214: Indians
Red Sox at Rays: 14.581, 15.103: Rays
Blue Jays at Braves: 12.117, 13.754: Braves
White Sox at Brewers: 14.154, 15.064: White Sox
Royals at Cubs: N/A
Giants at Rockies: 17.543, 14.124: Rockies
Dodgers at Padres: 16.31, 14.08: Dodgers
Astros at D-backs: N/A
Rangers at Athletics: 13.326, 16.179: Athletics
Mets at Nationals: 15.548, 16.149: Nationals
Angels at Mariners: 16.932, 10.294: Angels

2020-08-06
Giants at Rockies: 15.315, 14.156: Rockies
Angels at Mariners: 14.808, 11.903: Angels
Yankees at Phillies: 14.111, 15.354: Phillies
Reds at Indians: 14.173, 17.034: Indians

scripts/read_todays_games.py
import requests
from bs4 import BeautifulSoup
from simulate_game import simulate_games
import time
from datetime import date, datetime, timedelta
import json

def read_games(game_day_dict):
    URL = 'https://www.mlb.com/starting-lineups'
    #URL = 'https://www.mlb.com/starting-lineups/2020-07-26'
    page = requests.get(URL)

    soup = BeautifulSoup(page.content, 'html.parser')

    results = soup.find(id='starting-lineups_index')

    #print(results.prettify())

    job_elems = results.find_all('div', class_='starting-lineups__matchup')

    for job_elem in job_elems:
        game_identifier = job_elem['data-gamepk']
        if game_identifier in game_day_dict.keys():
            game_dict = game_day_dict[game_identifier]
        else:
            game_dict = {}
        if 'status' not in game_dict.keys():
            game_dict['status'] = 'game created'
        game_time = job_elem.find('div', class_='starting-lineups__game-date-time').text.strip()
        home_team = job_elem.find('span', class_='starting-lineups__team-name starting-lineups__team-name--home').text.strip()
        away_team = job_elem.find('span', class_='starting-lineups__team-name starting-lineups__team-name--away').text.strip()
        game_dict['id'] = game_identifier
        game_dict['home_team'] = home_team
        game_dict['away_team'] = away_team
        game_dict['date'] = game_time
        pitchers = job_elem.find_all('div', class_='starting-lineups__pitcher-name')
        ind = 0
        for pitcher in pitchers:
            if ind == 0:
                away_pitcher = pitcher.text.strip()
                ind = ind + 1
            else:
                home_pitcher = pitcher.text.strip()
        game_dict['home_pitcher'] = home_pitcher
        game_dict['away_pitcher'] = away_pitcher
        away_lineup = []
        home_lineup = []
        lineups = job_elem.find('div', class_='starting-lineups__teams starting-lineups__teams--sm starting-lineups__teams--xl')
        players = lineups.find_all('li', class_='starting-lineups__player')
        count = 0
        for player in players:
            player_stripped = player.text.strip()
            name = player_stripped.split(" ")
            if count < 9:
                if len(name) == 5:
                    away_lineup.append(name[0] + " " + name[1] + " " + name[2])
                else:
                    away_lineup.append(name[0] + " " + name[1])
            else:
                if len(name) == 5:
                    home_lineup.append(name[0] + " " + name[1] + " " + name[2])
                else:
                    home_lineup.append(name[0] + " " + name[1])
            count = count + 1
        game_dict['home_lineup'] = home_lineup
        game_dict['away_lineup'] = away_lineup

        if away_lineup and home_lineup and game_dict['status'] not in 'processed':
            simulate_games(away_pitcher, home_pitcher, away_team, home_team, away_lineup, home_lineup, game_dict)
            game_dict['status'] = 'processed'
        elif 'PPD' in game_time:
            game_dict['date'] = 'PPD'
            game_dict['status'] = 'processed'
        elif not away_lineup or not home_lineup:
            game_dict['status'] = 'waiting for lineups'

        game_day_dict[game_identifier] = game_dict
    return game_day_dict

if __name__ == "__main__":
    starttime = time.time()
    today = date.today()
    game_day_dict = {'631529': {'status': 'processed', 'id': '631529', 'home_team': 'Rockies', 'away_team': 'Giants', 'date': '3:10 PM', 'home_pitcher': 'Kyle Freeland', 'away_pitcher': 'Logan Webb', 'home_lineup': ['Raimel Tapia', 'Trevor Story', 'Nolan Arenado', 'Charlie Blackmon', 'Kevin Pillar', 'Garrett Hampson', 'Sam Hilliard', 'Josh Fuentes', 'Tony Wolters'], 'away_lineup': ['Donovan Solano', 'Mike Yastrzemski', 'Evan Longoria', 'Wilmer Flores', 'Darin Ruf', 'Alex Dickerson', 'Joey Bart', 'Daniel Robertson', 'Mauricio Dubon'], 'home_hits': '13.619', 'away_hits': '15.055'}, '631673': {'status': 'processed', 'id': '631673', 'home_team': 'Orioles', 'away_team': 'Mets', 'date': '4:05 PM', 'home_pitcher': 'John Means', 'away_pitcher': 'Michael Wacha', 'home_lineup': ['Cedric Mullins', 'Anthony Santander', 'Jose Iglesias', 'Renato Nunez', 'Pedro Severino', 'Ryan Mountcastle', 'Rio Ruiz', 'Pat Valaika', 'Andrew Velazquez'], 'away_lineup': ['Jeff McNeil', 'J.D. Davis', 'Michael Conforto', 'Todd Frazier', 'Robinson Cano', 'Pete Alonso', 'Wilson Ramos', 'Jake Marisnick', 'Amed Rosario'], 'home_hits': '9.87', 'away_hits': '15.371'}, '631587': {'status': 'processed', 'id': '631587', 'home_team': 'Reds', 'away_team': 'Cardinals', 'date': 'warmup\n                                                                                \n                                                                                                •', 'home_pitcher': 'Tyler Mahle', 'away_pitcher': 'Johan Oviedo', 'home_lineup': ['Joey Votto', 'Nick Castellanos', 'Jesse Winker', 'Eugenio Suarez', 'Mike Moustakas', 'Brian Goodwin', 'Jose Garcia', 'Shogo Akiyama', 'Tucker Barnhart'], 'away_lineup': ['Kolten Wong', 'Tommy Edman', 'Paul Goldschmidt', 'Brad Miller', 'Paul DeJong', 'Yadier Molina', 'Matt Carpenter', "Tyler O'Neill", 'Lane Thomas']}, '631308': {'status': 'processed', 'id': '631308', 'home_team': 'Marlins', 'away_team': 'Blue Jays', 'date': 'warmup\n                                                                                \n                                                                                                •', 'home_pitcher': 'Sixto Sanchez', 'away_pitcher': 'Hyun Jin Ryu', 'home_lineup': ['Jon Berti', 'Starling Marte', 'Garrett Cooper', 'Jesus Aguilar', 'Brian Anderson', 'Corey Dickerson', 'Lewis Brinson', 'Jorge Alfaro', 'Jazz Chisholm'], 'away_lineup': ['Cavan Biggio', 'Randal Grichuk', 'Jonathan Villar', 'Teoscar Hernandez', 'Rowdy Tellez', 'Lourdes Gurriel Jr.', 'Travis Shaw', 'Santiago Espinal', 'Danny Jansen']}, '631111': {'status': 'processed', 'id': '631111', 'home_team': 'Pirates', 'away_team': 'Cubs', 'date': '7:05 PM', 'home_pitcher': 'Joe Musgrove', 'away_pitcher': 'Kyle Hendricks', 'home_lineup': ['Cole Tucker', 'Adam Frazier', 'Kevin Newman', 'Josh Bell', 'Colin Moran', "Ke'Bryan Hayes", 'Gregory Polanco', 'Anthony Alford', 'Jacob Stallings'], 'away_lineup': ['Ian Happ', 'Willson Contreras', 'Anthony Rizzo', 'Javier Baez', 'Steven Souza Jr.', 'Jose Martinez', 'Cameron Maybin', 'David Bote', 'Nico Hoerner'], 'home_hits': '18.112', 'away_hits': '12.777'}, '631145': {'status': 'processed', 'id': '631145', 'home_team': 'Phillies', 'away_team': 'Nationals', 'date': '7:05 PM', 'home_pitcher': 'Zack Wheeler', 'away_pitcher': 'Max Scherzer', 'home_lineup': ['Andrew McCutchen', 'Didi Gregorius', 'Bryce Harper', 'J.T. Realmuto', 'Jay Bruce', 'Jean Segura', 'Alec Bohm', 'Neil Walker', 'Adam Haseley'], 'away_lineup': ['Trea Turner', 'Juan Soto', 'Howie Kendrick', 'Asdrubal Cabrera', 'Adam Eaton', 'Kurt Suzuki', 'Brock Holt', 'Luis Garcia', 'Victor Robles'], 'home_hits': '13.419', 'away_hits': '12.62'}, '631194': {'status': 'processed', 'id': '631194', 'home_team': 'Yankees', 'away_team': 'Rays', 'date': '7:05 PM', 'home_pitcher': 'Jordan Montgomery', 'away_pitcher': 'Charlie Morton', 'home_lineup': ['DJ LeMahieu', 'Luke Voit', 'Aaron Hicks', 'Clint Frazier', 'Gio Urshela', 'Mike Tauchman', 'Gary Sanchez', 'Brett Gardner', 'Tyler Wade'], 'away_lineup': ['Manuel Margot', 'Randy Arozarena', 'Austin Meadows', 'Mike Brosseau', 'Willy Adames', 'Joey Wendle', 'Hunter Renfroe', 'Kevan Smith', 'Nate Lowe'], 'home_hits': '15.181', 'away_hits': '17.078'}, '631624': {'status': 'processed', 'id': '631624', 'home_team': 'Red Sox', 'away_team': 'Braves', 'date': '7:30 PM', 'home_pitcher': 'Robinson Leyer', 'away_pitcher': 'Robbie Erlin', 'home_lineup': ['Alex Verdugo', 'Rafael Devers', 'Xander Bogaerts', 'J.D. Martinez', 'Kevin Plawecki', 'Michael Chavis', 'Jackie Bradley Jr.', 'Bobby Dalbec', 'Jose Peraza'], 'away_lineup': ['Dansby Swanson', 'Freddie Freeman', 'Marcell Ozuna', 'Nick Markakis', "Travis d'Arnaud", 'Austin Riley', 'Adam Duvall', 'Johan Camargo', 'Ender Inciarte']}, '631283': {'status': 'processed', 'id': '631283', 'home_team': 'Brewers', 'away_team': 'Tigers', 'date': '7:40 PM', 'home_pitcher': 'Adrian Houser', 'away_pitcher': 'Spencer Turnbull', 'home_lineup': ['Ben Gamel', 'Christian Yelich', 'Keston Hiura', 'Justin Smoak', 'Avisail Garcia', 'Omar Narvaez', 'Luis Urias', 'Eric Sogard', 'Orlando Arcia'], 'away_lineup': ['Victor Reyes', 'Jonathan Schoop', 'Miguel Cabrera', 'Jeimer Candelario', 'Willi Castro', 'Christin Stewart', 'Jorge Bonifacio', 'Austin Romine', 'Isaac Paredes'], 'home_hits': '15.217', 'away_hits': '13.972'}, '631404': {'status': 'processed', 'id': '631404', 'home_team': 'Royals', 'away_team': 'Indians', 'date': '8:05 PM', 'home_pitcher': 'Jakob Junis', 'away_pitcher': 'Triston McKenzie', 'home_lineup': ['Whit Merrifield', 'Hunter Dozier', 'Jorge Soler', 'Maikel Franco', 'Ryan McBroom', 'Alex Gordon', 'Adalberto Mondesi', 'Nicky Lopez', 'Cam Gallagher'], 'away_lineup': ['Cesar Hernandez', 'Jose Ramirez', 'Francisco Lindor', 'Carlos Santana', 'Franmil Reyes', 'Tyler Naquin', 'Josh Naylor', 'Roberto Perez', 'Delino DeShields']}, '631431': {'status': 'processed', 'id': '631431', 'home_team': 'Astros', 'away_team': 'Rangers', 'date': '8:10 PM', 'home_pitcher': 'Cristian Javier', 'away_pitcher': 'Kolby Allard', 'home_lineup': ['George Springer', 'Jose Altuve', 'Michael Brantley', 'Yuli Gurriel', 'Kyle Tucker', 'Carlos Correa', 'Josh Reddick', 'Aledmys Diaz', 'Martin Maldonado'], 'away_lineup': ['Leody Taveras', 'Isiah Kiner-Falefa', 'Shin-Soo Choo', 'Nick Solak', 'Joey Gallo', 'Elvis Andrus', 'Derek Dietrich', 'Anderson Tejeda', 'Jeff Mathis']}, '631253': {'status': 'processed', 'id': '631253', 'home_team': 'Twins', 'away_team': 'White Sox', 'date': '8:10 PM', 'home_pitcher': 'Jose Berrios', 'away_pitcher': 'Reynaldo Lopez', 'home_lineup': ['Max Kepler', 'Josh Donaldson', 'Jorge Polanco', 'Nelson Cruz', 'Eddie Rosario', 'Miguel Sano', 'Luis Arraez', 'Jake Cave', 'Ryan Jeffers'], 'away_lineup': ['Tim Anderson', 'Yoan Moncada', 'Yasmani Grandal', 'Jose Abreu', 'Eloy Jimenez', 'Edwin Encarnacion', 'Luis Robert', 'Nomar Mazara', 'Nick Madrigal'], 'home_hits': '14.293', 'away_hits': '14.268'}, '631356': {'status': 'processed', 'id': '631356', 'home_team': 'Dodgers', 'away_team': 'D-backs', 'date': '9:40 PM', 'home_pitcher': 'Walker Buehler', 'away_pitcher': 'Zac Gallen', 'home_lineup': ['Mookie Betts', 'Corey Seager', 'AJ Pollock', 'Max Muncy', 'Chris Taylor', 'Joc Pederson', 'Will Smith', 'Edwin Rios', 'Gavin Lux'], 'away_lineup': ['Kole Calhoun', 'Ketel Marte', 'Christian Walker', 'David Peralta', 'Eduardo Escobar', 'Josh Rojas', 'Nick Ahmed', 'Daulton Varsho', 'Carson Kelly'], 'home_hits': '15.935', 'away_hits': '14.22'}, '631379': {'status': 'waiting for lineups', 'id': '631379', 'home_team': 'Angels', 'away_team': 'Padres', 'date': '9:40 PM', 'home_pitcher': 'Julio Teheran', 'away_pitcher': 'Dinelson Lamet', 'home_lineup': [], 'away_lineup': []}, '631054': {'status': 'processed', 'id': '631054', 'home_team': 'Mariners', 'away_team': 'Athletics', 'date': 'PPD', 'home_pitcher': 'Yusei Kikuchi', 'away_pitcher': 'TBD', 'home_lineup': [], 'away_lineup': []}}
    still_running = True
    with open("../logs/history.txt", "a") as myfile:
        myfile.write("\n" + str(today) + "\n")
    while still_running:
        game_day_dict = read_games(game_day_dict)
        print(game_day_dict)
        still_running = False
        for item in game_day_dict.keys():
            if game_day_dict[item]['status'] not in 'processed':
                still_running = True
        if still_running:
            print("sleeping")
            timer = datetime.now() + timedelta(minutes=30)
            time_str = timer.strftime('%H:%M')
            print("next run at ", time_str)
            time.sleep(1800)
            print("done sleeping")
        #time.sleep(60.0 - ((time.time() - starttime) % 60.0))
    outfile_path = "../DateFiles/" + str(today) + ".json"
    with open(outfile_path, 'w', encoding='utf-8') as f:
        json.dump(game_day_dict, f, ensure_ascii=False, indent=4)
    print("************FINISHED************")
